working/macros/api/builders.dart

```diff
@@ -22,21 +22,24 @@ abstract class TypeBuilder implements Builder {
   void declareType(DeclarationCode typeDeclaration);
 }
 
+/// The interface for checking if a type implements another type.
+abstract class TypeComparator {
+  /// Returns true if [leftType] is a subtype of [rightType].
+  bool isSubtypeOf(TypeAnnotation leftType, TypeAnnotation rightType);
```

Comment:
How will clients use this? It looks that `TypeAnnotation` is practically just a name, a string. I can see how it can be used on one side - check what is written in the code. But if we want for example to know if something is a subtype of `Widget` from `package:flutter`, we need to tell that that `Widget` name *is* from `package:flutter`.

Reply:
+1. I really do think the API should make a clear distinction between a type annotation and a type.

Reply:
Yes I definitely agree this part of the api is funky. However, adding an api that takes a `TypeAnnotation` and resolves it just moves the problem to a different API.

One other option is to allow you to resolve entire `Declaration`s, which would resolve all the `TypeAnnotation`s in the declaration. But that has several downsides as well:

- It dramatically increases the api surface area (all declarations need resolved counterparts).
- It is less granular than we might want for many situations (I might only care about the return type of a function, etc).

My general thought is that these `TypeAnnotation` objects actually must originate from the implementation that is being asked this question - and thus they will actually be subtypes of `TypeAnnotation` that contain more information than just what `TypeAnnotation` has, and do have the ability to answer these questions.

It is funky though I agree for sure, because you can't express this in the API itself.

Would you all be OK with adding a TODO on this particular part for now, to get the rest landed, and then I can come back to this?

Reply:
Maybe the simplest solution here actually is to bring back the `Scope` object?

Reply:
OK, I see. So, `TypeAnnotation` could be anything - an unresolved annotation from AST, or an actual type requested in some way, like `typeProvider.getClassElement('package:flutter/widget.dart', 'Widget').instantiate([], noNullability)`.

One weird case if when both `left` and `right` type annotations are from AST. But I guess it still might be interesting sometimes.

Reply:
Ya, I don't envision macro authors building these on their own (and we may even want to try and prevent them from doing so?). So they always come from either an actual AST or `instantiate`, etc. We also want to have an api for getting one from a runtime type in the macro code itself at some point.

But conceptually a `TypeAnnotation` subtype should have more underlying information that is actually exposed through its own public api. It is difficult to come up with an agnostic way of representing that though. But if we brought back the opaque `Scope` class, it might help?

Reply:
I don't know what `Scope` is. Any `TypeAnnotation` from AST has its actual scope. And as much as I can imagine, the other type is requested as I described above - by the library URI and name.

One thing that is not clear to me, is if any AST type annotation is always resolved. During the phase one (when we can declare new types), we don't have types resolved yet, right? Then, what `TypeComparator` methods will return? `false`? Exception?

Reply:
> I don't know what `Scope` is.

It was intended as an opaque thing, that would only have a meaning to an individual implementation. It would hold the information required to know where a type annotation actually came from (which would be required to resolve it).

Ultimately I dropped it because it felt unnecessary (the implementations could just put this info on their own subtype).

It sounds to me like maybe that is still the right choice and adding the abstraction back as a user-visible but opaque thing might not really hold value (or worse, just be confusing).

> One thing that is not clear to me, is if any AST type annotation is always resolved.

They are not resolved, no. But they need to be _resolvable_, if that makes sense.

> During the phase one (when we can declare new types), we don't have types resolved yet, right? Then, what `TypeComparator` methods will return? `false`? Exception?

The phase one methods do not get a `TypeComparator`, their is no public api available to them to attempt to resolve things (and the executor function also doesn't accept one):

```dart
  Future<MacroExecutionResult> executeTypesPhase(
      MacroInstanceIdentifier macro, Declaration declaration);
```

Reply:
> One other option is to allow you to resolve entire `Declaration`s, which would resolve all the `TypeAnnotation`s in the declaration.

I don't think we need that. I think there are basically three concepts:

*   A **type annotation** corresponds to the actual syntax of a type annotation as it might appear in a program. Think a return type on a function, type on a variable declaration, etc. The identifiers in it might be resolved so that we know where they come from, but the object represents the *annotation*, not the type. For example:

    ```dart
    typedef AlsoInt = int;

    main() {
      int a;
      AlsoInt b;
    }
    ```

    Here, `a` and `b` have different type annotations, even though they end up meaning the same type.

*   A **type** is the entity that the static type system works with. It's the construct for which questions like "is this a subtype?" are meaningful. (Questions like this are not meaningful for type *annotations* because things like aliases obscure it.) Types are the inputs and outputs of type inference.

*   A **type declaration** is some Dart language construct that brings a new named type into being. Class declarations, enum declarations, mixin declarations. In some cases, you can think of those declarations as *being* their corresponding types, but it's a clearer model to separate them. In particular, with generic classes, there is a real distinction between the class declaration and the types around it. Given:

    ```dart
    class Box<T> {}
    ```

    The type `Box<int>` is a real type that is meaningful, but that type is not simply a pointer to the class declaration because it also tracks the set of type arguments applied to the class.

You can move between these three concepts and they overlap in some ways, but they're distinct. You can take a type annotation and "resolve" it get the type that it describes. If that type is derived from some nominal declaration like a class, you can probably get from the type to the class. But we can't collapse those last two categories *in general*. There are types like function types where there is no declaration to go to at all (also, eventually record types). And even instantiations of generic types need to also track their type arguments.

If I remember right, the analyzer element API doesn't model these three concepts clearly. Every type implicitly tries to also be a declaration. If I remember right, this is an artifact of the analyzer being written back before Dart had real support for function types. (All it had was function-typed parameters.) So, back then, any type you could talk about also had a declaration you could point to. But that's no longer the case.

I think the API we should have for introspection should treat them separately:

*   A base `TypeAnnotation` class for anything that can appear in a type annotation with subclasses for `NamedTypeAnnotation` and `FunctionTypeAnnotation`. (Maybe `NullableTypeAnnotation` if you want to consider nullable types a wrapper around their base type.) The properties of these are also type annotations.

    You can resolve one of these to get...

*   A base, uh, `StaticType` class for an object that represents a static type. These are the things you can ask about subtypes, same type, etc. Again subclasses for `NamedType` (which has a potentially empty list of type arguments) and `FunctionType`. The properties of these are `StaticTypes`.

    The `NamedType` class then has a property or operation to get to the declaration that introduces that name. That takes you to...

*   `ClassDeclaration`, `EnumDeclaration`, etc. These are essentially syntactic classes. They let you introspect over the structure and contents of the type. They will have a list of type parameters, superclass, superinterfaces, fields, etc. References to types in all of those are `TypeAnnotation`, again because this object represents a mostly-syntactic entity.

    In principle, sure, we could also offer something like a `ResolvedClassDeclaration` where every property that was a type annotation has been resolved to its corresponding type, but I don't think that's necessary or very useful.

Type aliases are the funny one. We could also have a operation on `NamedTypeAnnotation` to take you to the declaration where that name is defined. If you do that for a type annotation that refers to a type alias, it would take you to the `typedef` declaration, not the actual declaration of the underlying base type. 

Does this make sense?

Reply:
Actually, it was my personal crusade (one of :-)) to make function types completely synthetic. So, now `FunctionTypeImpl` does not know if it was created from a method, or specified as a generic function type. For code generation purposes (e.g. Angular) however we track if a type (any, not only function type) was create from a type alias.

What you describe about separation between declarations, types, and annotations, all makes sense to me, because this is how it is done in the analyzer.

For type annotations that reference a type alias `typedef MyAlias<T> = List<T>;`, like `MyAlias<int>` the name `MyAlias` references `MyAlias` in `typedef`. But the type is `List<int>`, somewhere in `dart:core`.

Reply:
I think this all almost basically matches the existing setup, but with the addition of a `StaticType` class.

Concretely we would change the `TypeIntrospector` api from this:

```dart
/// The api used by [Macro] to get a [TypeDeclaration] for any given
/// [NamedTypeAnnotation].
abstract class TypeIntrospector {
  /// Resolves a [NamedTypeAnnotation] to its declaration.
  Future<TypeDeclaration> resolve(NamedTypeAnnotation annotation);
}
```

To this:

```dart
/// The api used by [Macro] to get a [TypeDeclaration] for any given
/// [NamedTypeAnnotation].
abstract class TypeIntrospector {
  /// Resolves a [TypeAnnotation] to a [StaticType].
  Future<StaticType> resolve(TypeAnnotation annotation);

  /// Resolves a [NamedType] to its declaration.
  Future<TypeDeclaration> declarationOf(NamedType annotation);
}
```

And then move the `isSubtypeOf` and `isExactly` apis from the `TypeComparator` interface to this new `StaticType` interface.

I do think it makes sense to add also a `TypeAliasDeclaration` class, and give back that when asking for the declaration of a named type that is actually an alias (and that will have a `TypeAnnotation get type;` to get the actual type).

## Pros

- There is a more explicit line drawn between a type annotation and a type.

## Cons

- It requires 2 async operations in order to go from a `TypeAnnotation` to its declaration, instead of one in the current api.
  -  We could also have an api to go from a `TypeAnnotation` directly to the declaration to avoid this if desired (ie: the current api).
- It still has all the same implicit problems of the existing api where `TypeAnnotation`s are a bit weird, and must have additional implementation specific tracking information in them in order to be resolvable. So I don't think it concretely changes anything, but maybe produces a more explicit api that is easier to understand?

I am

Reply:
LGTM!

Reply:
Ok I pushed up some of these changes, ended up a bit more complex than I expected but the gist is the same. There is a bit to still work out in terms of how these StaticTypes can actually be converted to (or used by) Code objects, so we can build up a proper augmentation library from them etc.

But lets work that out separately.